### PR TITLE
8271398: GTK3 drag view image swaps red and blue color channels

### PR DESCRIPTION
**Pull request by tsayao**
It seems raw images need to be converted BRGA -> RGBA.

It was being converted on gtk2 code path, but gtk3 only uses `gtk_drag_set_icon_pixbuf`.

I have simplified the gtk2 `DragView::View::expose` to paint with `gdk_cairo_set_source_pixbuf` (that is available since Gtk 2.8) because the old way was somehow converting  again.

Run the issue sample with `-Djdk.gtk.version=2` to test the gtk2 code path.

To test:

`./gradlew apps`

```
java @build/run.args -cp apps/toys/DragDrop/dist/DragDrop.jar dragdrop.DragDropWithControls
java @build/run.args -cp apps/toys/DragDrop/dist/DragDrop.jar dragdrop.DragDropColor 

java -Djdk.gtk.version=2 @build/run.args -cp apps/toys/DragDrop/dist/DragDrop.jar dragdrop.DragDropWithControls
java -Djdk.gtk.version=2 @build/run.args -cp apps/toys/DragDrop/dist/DragDrop.jar dragdrop.DragDropColor 
```

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8271398](https://bugs.openjdk.java.net/browse/JDK-8271398): GTK3 drag view image swaps red and blue color channels

### Reviewers
 * [Pankaj Bansal](https://openjdk.java.net/census#pbansal) (@pankaj-bansal - Committer)
 * [Kevin Rushforth](https://openjdk.java.net/census#kcr) (@kevinrushforth - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jfx pull/599/head:pull/599` \
`$ git checkout pull/599`

Update a local copy of the PR: \
`$ git checkout pull/599` \
`$ git pull https://git.openjdk.java.net/jfx pull/599/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 599`

View PR using the GUI difftool: \
`$ git pr show -t 599`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jfx/pull/599.diff">https://git.openjdk.java.net/jfx/pull/599.diff</a>

</details>

### CODE REVIEW

**Comment by pankaj-bansal** [comment]
I will look at this. Meanwhile, could you please write an automated system test for this?

**Comment by tsayao** [comment]
> I will look at this. Meanwhile, could you please write an automated system test for this?

Sure, I would provide it, but in the past drag and drop tests were not possible. Any ideas?

**Comment by pankaj-bansal** [comment]
> > I will look at this. Meanwhile, could you please write an automated system test for this?
> 
> Sure, I would provide it, but in the past drag and drop tests were not possible. Any ideas?

I think it should be possible to write an automated testcase in this case. Find center of window, move mouse to the location and do mouse click (without release). Then move mouse out of window to some location. Afterwards, get the pixel color at mouse location and do mouse release. Looks like it should be possible. Please let me know if there is some problem, I will also try to write one on my end.

**Comment by tsayao** [comment]
Test works on Linux, don't know on other platforms.

`gradlew -PFULL_TEST=true -PUSE_ROBOT=true :systemTests:test --tests test.robot.javafx.dnd.DndRawImageTest`

**Comment by pankaj-bansal** [comment]
> Test works on Linux, don't know on other platforms.
> 
> `gradlew -PFULL_TEST=true -PUSE_ROBOT=true :systemTests:test --tests test.robot.javafx.dnd.DndRawImageTest`

The fix does solve the issue and looks good to me.

The test is failing on Windows 10 and Mac 10.15. As Kevin has mentioned in other comment, if the automated test is too complex, you can change the test to manual one and file a new bug for automated test and assign it to me. I will write an automated in future.

**Comment by tsayao** [comment]
@pankaj-bansal Do you prefer I switch the test to manual or make the current one only for Linux?

**Comment by pankaj-bansal** [comment]
> @pankaj-bansal Do you prefer I switch the test to manual or make the current one only for Linux?

I think it would be better to make it manual and create new JBS bug for automated test.

**Comment by tsayao** [comment]
JBS: https://bugs.openjdk.java.net/browse/JDK-8272492
Changed test to manual.

**Comment by tsayao** [comment]
`gdk_cairo_set_source_pixbuf' is available since gtk 2.8. I can't find the minimum gtk version for 2.0.

File: tests/manual/dnd/DndTestDragViewRawImage.java [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+                } else if (x < width * 0.66) {
+                    image.setRGB(x, y, 0x00FF00);
+                } else {
+                    image.setRGB(x, y, 0x0000FF);
+                }
+            }
+        }
+        return SwingFXUtils.toFXImage(image, null);
+    }
+}
```
Comment (pankaj-bansal) [inline tests/manual/dnd/DndTestDragViewRawImage.java:83]: There is no newline at the file end

File: tests/manual/dnd/DndTestDragViewRawImage.java [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+        ImageView imageView = new ImageView(image);
+        imageView.setOnDragDetected(event -> {
+            ClipboardContent content = new ClipboardContent();
+            content.putImage(image);
+            Dragboard dragboard = imageView.startDragAndDrop(TransferMode.ANY);
+            dragboard.setContent(content);
+            dragboard.setDragView(image);
+        });
+
+        Label label = new Label("Drag image should match source colors when dragged");
```
Comment (pankaj-bansal) [inline tests/manual/dnd/DndTestDragViewRawImage.java:59]: Could you please make the instructions a bit more clear?

**Comment by pankaj-bansal** [comment]
The fix works fine and the test passes on all platforms. I have given minor comments about the test.

File: tests/manual/dnd/DndTestDragViewRawImage.java (hunk 2 above)
Comment (tsayao) [reply tests/manual/dnd/DndTestDragViewRawImage.java]: I tried :)

**Review by pankaj-bansal — APPROVED** [review]
Looks good to me now

**Review by kevinrushforth — COMMENTED** [review]
This fix introduces a memory leak, but otherwise looks correct. I verified that with gtk3 the new manual test passes with the fix and fails without the fix (with gtk2 is passes before and after the fix).

File: modules/javafx.graphics/src/main/native-glass/gtk/glass_dnd.cpp [hunk 3]
```diff
@@ -911,6 +911,11 @@ GdkPixbuf* DragView::get_drag_image(GtkWidget *widget, gboolean* is_raw_image, g
                     guchar* data = (guchar*) g_try_malloc0(nraw - whsz);
                     if (data) {
                         memcpy(data, (raw + whsz), nraw - whsz);
+
+                        if (is_raw_image) {
+                            data = (guchar*) convert_BGRA_to_RGBA((const int*) data, w * 4, h);
```
[the remark this answers is not in the document]
Comment (tsayao) [reply modules/javafx.graphics/src/main/native-glass/gtk/glass_dnd.cpp:917]: Fixed. I really felt something was missing :)

**Review by kevinrushforth — APPROVED** [review]

**Review by pankaj-bansal — APPROVED** [review]
Looks good to me. I re-tested the test with both gtk2 and gtk3 and it the issue is fixed.

**Pull request closed without merging**